docs/reference/engine-config.rst

.. _engine_config_options:

SCQLEngine Configuration
------------------------

SCQLEngine uses Gflags to manage configurations when SCQLEngine set up.

Example configuration for SCQLEngine
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block::

  # Config for Brpc server
  --listen_port=8003
  # Config for datasource
  --datasource_router=embed
  --embed_router_conf={"datasources":[{"id":"ds001","name":"mysql db","kind":"MYSQL","connection_str":"${connection_str}"}],"rules":[{"db":"*","table":"*","datasource_id":"ds001"}]}


Configuration Options of SCQLEngine
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

SCQLEngine works with the native compiler to execute query plans. In the configuration, ``Driver`` refers to the module that sends execution plans to the engine.

+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
|                      Name                       |       Default       |                                                               Description                                                               |
+=================================================+=====================+=========================================================================================================================================+
| log_enable_console_logger                       | true                | Whether logging to stdout while logging to file                                                                                         |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| log_enable_session_logger_separation            | false               | Whether output session-related logs to a dedicated file                                                                                 |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| log_dir                                         | logs                | The directory to save log file                                                                                                          |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| log_level                                       | info                | Log level, can be trace/debug/info/warning/error/critical/off                                                                           |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_protocol                            | baidu_std           | The rpc protocol between engine and engine                                                                                              |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_connection_type                     | single              | The rpc connection type between engine and engine                                                                                       |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_load_balancer                       | none                | The rpc load balancer between engine and engine, can be rr or empty string                                                              |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_timeout_ms                          | 300000              | The rpc timeout between engine and engine, unit: ms                                                                                     |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_max_retry                           | 3                   | Rpc max retries(not including the first rpc) between engine and engine                                                                  |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_enable_ssl_as_client                | true                | Whether enable ssl encryption when send message to another engine                                                                       |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_enable_ssl_client_verification      | false               | Whether enable certificate verification when send message to another engine                                                             |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| peer_engine_ssl_client_ca_certificate           | none                | The trusted CA file to verify certificate when send message to another engine                                                           |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| link_recv_timeout_ms                            | 30000               | The max time that engine will wait for message come from another engine                                                                 |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| link_throttle_window_size                       | 16                  | Throttle window size for channel, set to limit the number of messages sent asynchronously to avoid network congestion, set 0 to disable |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| link_chunked_send_parallel_size                 | 1                   | Parallel size when send chunked value                                                                                                   |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| http_max_payload_size                           | 1048576             | Max payload to decide whether to send value chunked, default 1MB                                                                        |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_protocol                                 | `http:proto`        | The rpc protocol between engine and Driver                                                                                              |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_connection_type                          | pooled              | The rpc connection type between engine and Driver                                                                                       |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_load_balancer                            | none                | The rpc load balancer between engine and Driver, can be rr or empty string                                                              |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_timeout_ms                               | 5000                | The rpc timeout between engine and Driver, unit: ms                                                                                     |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_max_retry                                | 3                   | Rpc max retries(not including the first rpc) between engine and Driver                                                                  |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_enable_ssl_as_client                     | true                | Whether enable ssl encryption when send message to Driver                                                                               |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_enable_ssl_client_verification           | false               | Whether enable certificate verification when send message to Driver                                                                     |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| driver_ssl_client_ca_certificate                | none                | The trusted CA file to verify certificate when send message to Driver                                                                   |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| listen_port                                     | 8003                | The listening port of engine service                                                                                                    |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| enable_builtin_service                          | false               | Whether enable brpc builtin service                                                                                                     |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| internal_port                                   | 9527                | The listening port of brpc builtin services                                                                                             |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| enable_separate_link_port                       | false               | Whether use a separate port for link service                                                                                            |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| link_port                                       | 8004                | Port for link service                                                                                                                   |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| idle_timeout_s                                  | 30                  | Idle connection close delay in seconds between the engine and Driver, unit: s                                                           |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| server_enable_ssl                               | true                | Whether enable SSL when engine work as a server                                                                                         |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| server_ssl_certificate                          | none                | Certificate file path to enable SSL when engine work as a server                                                                        |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| server_ssl_private_key                          | none                | Private key file path to enable SSL when engine work as a server                                                                        |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| enable_client_authorization                     | false               | Whether check requests' http header when engine work as a server                                                                        |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| auth_credential                                 | none                | Authorization credential used to check requests' http header                                                                            |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| enable_driver_authorization                     | false               | Whether to authenticate the identity of Driver                                                                                          |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| engine_credential                               | none                | Credential used to authenticate Driver                                                                                                  |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| session_timeout_s                               | 1800                | Expiration duration of a session between engine and Driver, unit: s                                                                     |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| spu_allowed_protocols                           | SEMI2K,ABY3,CHEETAH | SPU allowed protocols                                                                                                                   |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| datasource_router                               | embed               | The datasource router type, "embed" or "http"                                                                                           |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| embed_router_conf                               | none                | Configuration for embed router in json format                                                                                           |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| http_router_endpoint                            | none                | http datasource router endpoint, it is valid only datasource_router is set to "http"                                                    |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| kuscia_datamesh_endpoint                        | datamesh:8071       | Kuscia datamesh grpc endpoint                                                                                                           |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| kuscia_datamesh_client_key_path                 | none                | Kuscia datamesh client key file                                                                                                         |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| kuscia_datamesh_client_cert_path                | none                | Kuscia datamesh client cert file                                                                                                        |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| kuscia_datamesh_cacert_path                     | none                | Kuscia datamesh server cacert file                                                                                                      |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| db_connection_info                              | none                | Connection string used to connect to mysql                                                                                              |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| enable_self_auth                                | true                | Whether enable self identity authentication                                                                                             |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| private_key_pem_path                            | none                | Path to private key pem file                                                                                                            |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| enable_peer_auth                                | true                | Whether enable peer parties identity authentication                                                                                     |
+-------------------------------------------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------+
| authorized_profile_path                         | none                | Path to authorized profile, in json format                                                                                              |

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Types of changes

`Added` for new features.
`Changed` for changes in existing functionality.
`Deprecated` for soon-to-be removed features.
`Removed` for now removed features.
`Fixed` for any bug fixes.
`Security` in case of vulnerabilities.

## Staging

### Added

### Changed

### Fixed

## [0.9.4] - 2025-07-29

### Added

- Added Archive API to support project archiving.
- Enhanced time-type data processing capability: support `STR_TO_DATE` function and implicit conversion from string to time types.
- Supported richer expression of statements: `PERCENTILE_DISC`, `BETWEEN AND`, `REPLACE`, etc.

### Changed

- Optimized data source reading, improving streaming processing capabilities.
- Optimized `JOIN` process to eliminate the need for additional `PLAINTEXT_AFTER_JOIN` CCL for non-result-receiving parties on join keys.

### Fixed

- Resolved the issue with compare subquery exceptions in aggregation scenarios.
- Fixed column disorder issues in the project tables.
- Resolved problems in LogicalOptimizer to prevent the removal of LogicalProjection nodes that could lead to performance and Tensor property inference issues.

## [0.9.3] - 2025-03-07

### Added

- Support for datasource `Doris 2.1.7`.
- Support `PERCENT_RANK` window function.
- Support various string-related single-party operators, including `UPPER`, `LOWER`, `SUBSTRING`, `TRIM`, `CONCAT` and others.
- Support `Scalar Subquery`, the subquery in the right is scalar value, e.g. SELECT * FROM ta JOIN tb ON ta.ID = tb.ID WHERE ta.salary > (SELECT AVG(ta.salary) FROM ta).
- Support `Compare Subquery`, allows comparison with ANY or ALL of the subquery results, e.g. SELECT * FROM ta JOIN tb ON ta.ID = tb.ID WHERE ta.salary > ANY(SELECT ta.salary FROM ta), However, comparisons using = or != are not supported in the HAVING clause. For instance, HAVING SUM(ta.salary) = ANY(SELECT salary FROM ta) is not supported.

### Changed

- Improved `JOIN` and `IN` performance in streaming mode.
- Implemented a more reliable `secret join algorithm`(only works in SEMI2K protocol) inspired by [Scape](https://ieeexplore.ieee.org/document/9835540/).
- Optimized the column pruning rule for Join, Selection, and Window nodes in the Logical Optimizer to more effectively remove redundant columns.

### Fixed

- Restricted access to SCQLEngine metrics using additional paths like "engine_ip:engine_port/metrics/additional/path".
- Prevented creation of tables with the same ref_table name but different db_type
- Fixed job creation error when selecting 'OPRF-PSI' but 'server hint' was missing.

## [0.9.2] - 2024-12-23

### Added

- Enhancement: Support `JOIN` after `UNION` operation.
- Add SCQL Agent to facilitate running SCQL query tasks in Kuscia, making it easier to integrate into SecretPad.
- Support writing results into multi-parties via `SELECT INTO OUTFILE` syntax.
- Support datasource `ODPS` via integrating with [dataproxy](https://github.com/secretflow/dataproxy).
- Support `order by`.
- Support a lot of single-party operators, such as `ABS`, `ASIN`, `EXP`, `FLOOR`, `SQRT` etc.

### Changed

- Improve the `JOIN` and `IN` performance via integrating [RR22 PSI](https://github.com/secretflow/psi/blob/v0.5.0b0/psi/proto/psi_v2.proto#L62).
- Improve the aggregation with group by performance if `reveal_group_count` enabled.

### Fixed

- Fixed an occasional crash issue when canceling query job.
- Fixed `select now()` is not supported issue.

## [0.9.1] - 2024-10-16

### Added

- Support window function `ROW_NUMBER()` with PARTITION BY clause and ORDER BY clause.
- Add new CCL constraint `REVAL_RANK`.
- Add ExplainQuery API with path `/intra/query/explain`.
- Support `INSERT INTO SELECT` syntax to allow writing query result back to db (mysql/sqlite/postgres).
- Support `trim` function.

### Changed

- Improved the job watcher to work better in broker clustered mode.

## [0.9.0] - 2024-08-01

### Added

- Support write outfile to OSS/MINIO via `select into` query.
- Support `sin`, `cos`, `acos` function.
- Support `geodist` function.
- Broker support using postgres as metadata storage.

### Changed

- Reduce the memory peak of large-scale intersection tasks through streaming execution.
- Link tcmalloc to solve the problem of memory increase.

### Fixed

- Fix crashes when dumpfile exceeds 2GB string column.
- Reduce the probability of graph checksum inconsistency issues.

## [0.8.1] - 2024-07-02

### Added

- Support session-based log isolation functioality in the SCQL Engine.
- Support consul-based broker registration/discovery services, providing ACL/TLS authentication.

### Changed

### Fixed

## [0.8.0] - 2024-06-12

### Added

- Enhanced `FetchResult` RPC and `brokerctl get result` command to report job progress when result is not ready.
- Support project/query level configs
- Support NULL for private data: including Arithmetic, Logic, Aggregation, etc., {IS [NOT] NULL, IFNULL, COALESCE} are also supported.
- Support port isolation for engine link service and control panel service (RunExecutionPlan).
- Add new CCL constraint `PLAINTEXT_AS_JOIN_PAYLOAD`.

### Changed

- **breaking**: The response value type of Broker API `DoQuery` and `FetchResult` have incompatible changes.

### Fixed

## [0.7.0] - 2024-05-14

### Added

- Added CheckAndUpdate API for self-recovery when status is inconsistent in P2P mode.

### Fixed

- Fixed the problem that Broker was unable to detect SCQLEngine crashes or being killed by OOM.

## [0.6.0] - 2024-04-15

### Added

- Support for RSA key pairs in SCQLBroker.
- Support running on [kuscia](https://github.com/secretflow/kuscia) and scheduling SCQLEngine dynamic via kuscia job.

engine/exe/flags.cc
// Copyright 2024 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "gflags/gflags.h"

// Log flags
DEFINE_string(
    log_level, "info",
    "log level, can be trace, debug, info, warning, error, critical, off");
DEFINE_string(log_dir, "logs", "log directory");
DEFINE_bool(log_enable_console_logger, true,
            "whether logging to stdout while logging to file");
DEFINE_bool(log_enable_session_logger_separation, false,
            "whether output session-related logs to a dedicated file");

// Brpc channel flags for peer engine.
DEFINE_string(peer_engine_protocol, "baidu_std", "rpc protocol");
DEFINE_string(peer_engine_connection_type, "single", "connection type");
DEFINE_string(peer_engine_load_balancer, "", "load balancer: \"rr\" or empty");
DEFINE_int32(peer_engine_timeout_ms, 300000, "rpc timeout in milliseconds.");
DEFINE_int32(peer_engine_max_retry, 3,
             "rpc max retries(not including the first rpc)");
DEFINE_bool(peer_engine_enable_ssl_as_client, true,
            "enable ssl encryption as client");
DEFINE_bool(peer_engine_enable_ssl_client_verification, false,
            "enable ssl client verification");
DEFINE_string(peer_engine_ssl_client_ca_certificate, "",
              "certificate Authority file path to verify SSL as client");
DEFINE_int32(link_recv_timeout_ms, 30 * 1000,
             "the max time that a party will wait for a given event");
DEFINE_int32(
    link_throttle_window_size, 16,
    "throttle window size for channel, set to limit the number of messages "
    "sent asynchronously to avoid network congestion, set 0 to disable");
// TODO(zhihe): for rpc stabilization in 100M bandwidth, the default
// link_chunked_send_parallel_size is 1, should find a tradeoff value between
// bandwidth and send speed
DEFINE_int32(link_chunked_send_parallel_size, 1,
             "parallel size when send chunked value");
DEFINE_int32(http_max_payload_size, 1024 * 1024,
             "max payload to decide whether to send value chunked, default 1M");
// Brpc channel flags for driver(for example, SCQLBroker)
DEFINE_string(driver_protocol, "http:proto", "rpc protocol");
DEFINE_string(driver_connection_type, "pooled", "connection type");
DEFINE_string(driver_load_balancer, "", "load balancer: \"rr\" or empty");
DEFINE_int32(driver_timeout_ms, 5000, "rpc timeout in milliseconds.");
DEFINE_int32(driver_max_retry, 3,
             "rpc max retries(not including the first rpc)");
DEFINE_bool(driver_enable_ssl_as_client, true,
            "enable ssl encryption as client");
DEFINE_bool(driver_enable_ssl_client_verification, false,
            "enable ssl client verification");
DEFINE_string(driver_ssl_client_ca_certificate, "",
              "certificate Authority file path to verify SSL as client");
// Brpc server flags.
DEFINE_int32(listen_port, 8003, "");
DEFINE_bool(enable_builtin_service, false,
            "whether brpc builtin service is enable/disable");
DEFINE_int32(internal_port, 9527, "which port the builtin services server on");
DEFINE_bool(enable_separate_link_port, false,
            "whether use a separate port for link service");
DEFINE_int32(link_port, 8004, "port for link service");
DEFINE_int32(idle_timeout_s, 30, "connections idle close delay in seconds");
DEFINE_bool(server_enable_ssl, true,
            "whether brpc server's ssl enable/disable");
DEFINE_bool(server_enable_ssl_client_verification, false,
            "enable ssl client verification");
DEFINE_string(server_ssl_certificate, "",
              "Certificate file path to enable SSL");
DEFINE_string(server_ssl_private_key, "",
              "Private key file path to enable SSL");
DEFINE_string(server_ssl_client_ca_certificate, "",
              "the trusted CA file to verity the client's certificate");
// Common flags for Brpc server and channel of peer engine.
DEFINE_bool(enable_client_authorization, false,
            "if set true, server will check all requests' http header.");
DEFINE_string(auth_credential, "", "authorization credential");
DEFINE_bool(enable_driver_authorization, false,
            "if set to true, the engine will verify the HTTP header "
            "'credential' field of requests from driver");
DEFINE_string(engine_credential, "", "driver authorization credential");
// Session flags
DEFINE_int32(session_timeout_s, 1800,
             "TTL for session, should be greater than the typical runtime of "
             "the specific tasks.");
DEFINE_string(spu_allowed_protocols, "SEMI2K,ABY3,CHEETAH",
              "spu allowed protocols");

// Kpad task mode flags
DEFINE_string(kpad_job_id, "", "job identifier for kpad task mode execution");
DEFINE_string(kpad_scql_config, "",
              "SCQL configuration in JSON format for kpad task mode");
DEFINE_string(kpad_cluster_def, "",
              "cluster definition in JSON format for kpad task mode");

// DataSource connection flags.
DEFINE_string(datasource_router, "embed",
              "datasource router type: embed | http | kusciadatamesh");
DEFINE_string(
    embed_router_conf, "",
    R"text(configuration for embed router in json format. For example: --embed_router_conf={"datasources":[{"id":"ds001","name":"mysql db","kind":"MYSQL","connection_str":"host=127.0.0.1 db=test user=root password='qwerty'"}],"rules":[{"db":"*","table":"*","datasource_id":"ds001"}]} )text");
DEFINE_string(http_router_endpoint, "", "http datasource router endpoint url");
DEFINE_string(kuscia_datamesh_endpoint, "datamesh:8071",
              "kuscia datamesh grpc endpoint");
DEFINE_string(kuscia_datamesh_client_key_path, "",
              "kuscia datamesh client key file");
DEFINE_string(kuscia_datamesh_client_cert_path, "",
              "kuscia datamesh client cert file");
DEFINE_string(kuscia_datamesh_cacert_path, "",
              "kuscia datamesh server cacert file");
DEFINE_string(db_connection_info, "",
              "connection string used to connect to mysql...");
// Party authentication flags
DEFINE_bool(enable_self_auth, true,
            "whether enable self identity authentication");
DEFINE_string(private_key_pem_path, "", "path to private key pem file");
DEFINE_bool(enable_peer_auth, true,
            "whether enable peer parties identity authentication");
DEFINE_string(authorized_profile_path, "",
              "path to authorized profile, in json format");
DEFINE_bool(enable_psi_detail_logger, false, "whether enable detail log");
DEFINE_string(psi_detail_logger_dir, "logs/detail", "log dir");
DEFINE_bool(enable_trace, false, "whether enable tracing");
DEFINE_string(trace_log_path, "logs/trace.trace", "trace log path");
DEFINE_int32(rr22_mode, 2, "rr22 comm mode, use fast mode by default");
DEFINE_int32(queue_max_block_seconds, 300, "max seconds for queue block wait");
